DataStructures/Lists/CircularBuffer.cs

```diff
@@ -0,0 +1,87 @@
+﻿using System;
+
+namespace DataStructures.Lists {
+    public class CircularBuffer<T> {
+        /// <summary>
+        /// Initializes a circular buffer with initial length of 10
+        /// </summary>
+        public CircularBuffer(bool canOverride=true) : this(10,canOverride) {
+        }
+
+        /// <summary>
+        /// Initializes a circular buffer with given length
+        /// </summary>
+        /// <param name="length">The length of the buffer</param>
+        public CircularBuffer(int length, bool canOverride=true) {
+            if (length < 1) {
+                throw new ArgumentOutOfRangeException("length can not be zero or negative");
+            }
+            _circularBuffer = new T[length + 1];
+            _end = 0;
+            _start = 0;
+            CanOverride = canOverride;
+        }
+
+        /// <summary>
+        /// Writes value to the back of the buffer
+        /// </summary>
+        /// <param name="value">value to be added to the buffer</param>
+        public void Write(T value) {
+            if (CanOverride) {
+                InsertData(value);
+            }
+            else {
+                DontOverrides(value);
+            }
+        }
+
+        /// <summary>
+        /// Inserts data into the buffer when it is not filled up
+        /// </summary>
+        /// <param name="value"></param>
+        private void DontOverrides(T value) {
+            if (IsFilledUp) {
+                return;
+            }
+            InsertData(value);
+        }
+
+        /// <summary>
+        /// Inserts data into the buffer without checking if it is full
+        /// </summary>
+        /// <param name="value"></param>
+        private void InsertData(T value) {
+            _circularBuffer[_end] = value;
+            _end = (_end + 1) % _circularBuffer.Length;
+            if (_end == _start) {
+                _start = (_start + 1) % _circularBuffer.Length;
+            }
+        }
+
+        /// <summary>
+        ///     Reads value from the front of the buffer
+        ///     <summay>
+        ///         Use the IsEmpty property to know when to read appropriately.
+        ///     </summay>
+        /// </summary>
```

Comment:
What effect the extra inside summary block has? Can we safe remove that?

Reply:
Just figured that it doesn't have much to offer. I have removed it.